File: utils/career_matcher.py
def calculate_career_matches(user_profile):
    """Calculate career matches based on user profile"""
    from data.careers import careers
    
    matches = []
    
    for career in careers:
        score = 0
        
        # RIASEC matching (40% weight)
        primary_score = user_profile['riasecScores'].get(career['primary_type'], 0)
        secondary_score = user_profile['riasecScores'].get(career.get('secondary_type', ''), 0)
        riasec_score = (primary_score * 0.7 + secondary_score * 0.3) / 5 * 40
        
        # Skills matching (35% weight)
        skills_match = 0
        if user_profile['skillsConfidence']:
            matched_skills = 0
            for skill in career['required_skills']:
                if skill in user_profile['skillsConfidence']:
                    matched_skills += user_profile['skillsConfidence'][skill] / 5
            skills_score = (matched_skills / len(career['required_skills'])) * 35 if career['required_skills'] else 0
        else:
            skills_score = 0
        
        # Work values matching (25% weight)
        values_score = 0
        if user_profile['workValues']:
            value_keywords = {
                'Work-Life Balance': ['flexible', 'balance', 'remote'],
                'Job Security': ['stable', 'secure', 'established'],
                'High Earnings': ['high salary', 'lucrative', 'well-paid'],
                'Helping Others': ['help', 'serve', 'support', 'care'],
                'Creativity': ['creative', 'innovative', 'design'],
                'Leadership': ['lead', 'manage', 'direct'],
                'Continuous Learning': ['learn', 'grow', 'develop'],
                'Recognition': ['recognition', 'prestige', 'respected']
            }
            
            matched_values = 0
            for value in user_profile['workValues']:
                keywords = value_keywords.get(value, [])
                for keyword in keywords:
                    if any(keyword in env.lower() for env in career.get('work_environment', [])):
                        matched_values += 1
                        break
            
            values_score = (matched_values / len(user_profile['workValues'])) * 25 if user_profile['workValues'] else 0
        
        total_score = riasec_score + skills_score + values_score
        
        matches.append({
            **career,
            'matchScore': round(total_score)
        })
    
    # Sort by match score
    matches.sort(key=lambda x: x['matchScore'], reverse=True)
    
    return matches[:10]  # Return top 10 matches

File: src/utils/careerMatcher.ts
import { UserProfile, CareerPath, RIASECScore } from '../types';
import { careerPaths } from '../data/careers';

export const calculateCareerMatches = (userProfile: UserProfile): CareerPath[] => {
  const scoredCareers = careerPaths.map(career => {
    let matchScore = 0;
    
    // RIASEC type matching (40% weight)
    const primaryTypeScore = userProfile.riasecScores[career.primaryType] || 0;
    const secondaryTypeScore = career.secondaryType 
      ? userProfile.riasecScores[career.secondaryType] || 0 
      : 0;
    
    const riasecScore = (primaryTypeScore * 0.7 + secondaryTypeScore * 0.3);
    matchScore += riasecScore * 0.4;
    
    // Skills matching (35% weight)
    const userSkills = Object.keys(userProfile.skillsConfidence);
    const matchingSkills = career.requiredSkills.filter(skill => 
      userSkills.some(userSkill => 
        userSkill.toLowerCase().includes(skill.toLowerCase()) ||
        skill.toLowerCase().includes(userSkill.toLowerCase())
      )
    );
    
    const skillsScore = (matchingSkills.length / career.requiredSkills.length) * 100;
    matchScore += skillsScore * 0.35;
    
    // Work values alignment (25% weight)
    const valueKeywords = {
      'Work-Life Balance': ['flexible', 'balance', 'remote'],
      'High Salary': ['salary', 'compensation', 'financial'],
      'Job Security': ['stable', 'secure', 'established'],
      'Creative Freedom': ['creative', 'artistic', 'innovative'],
      'Helping Others': ['social', 'helping', 'service'],
      'Recognition': ['leadership', 'management', 'achievement'],
      'Autonomy': ['independent', 'self-directed', 'entrepreneurial'],
      'Intellectual Challenge': ['analytical', 'problem solving', 'research'],
      'Variety': ['diverse', 'varied', 'different'],
      'Advancement Opportunities': ['growth', 'career progression', 'leadership']
    };
    
    let valuesScore = 0;
    userProfile.workValues.forEach(value => {
      const keywords = valueKeywords[value as keyof typeof valueKeywords] || [];
      const careerText = `${career.description} ${career.workEnvironment.join(' ')}`.toLowerCase();
      
      if (keywords.some(keyword => careerText.includes(keyword))) {
        valuesScore += 20; // Each matching value adds 20 points
      }
    });
    
    valuesScore = Math.min(valuesScore, 100); // Cap at 100
    matchScore += valuesScore * 0.25;
    
    return {
      ...career,
      matchScore: Math.round(matchScore)
    };
  });
  
  // Sort by match score and return top matches
  return scoredCareers
    .sort((a, b) => (b.matchScore || 0) - (a.matchScore || 0))
    .slice(0, 6);
};
